Source/SteveSloth/Private/MyProjectileBaseClass.cpp

```diff
@@ -15,32 +15,213 @@ AMyProjectileBaseClass::AMyProjectileBaseClass()
+
+	Player = USteveSingleton::GetSteve()->GetPlayerCharacter();
+
+	if (IsValid(Player))
+	{
+		Steve = Cast<AMyPlayer>(Player);
+	}
+
+	StartingLocation = GetActorLocation();
+
+	ProjectileHitbox->OnComponentBeginOverlap.AddDynamic(this, &AMyProjectileBaseClass::OnHitboxOverlapBegin);
+	AreaOfEffectHitbox->OnComponentBeginOverlap.AddDynamic(this, &AMyProjectileBaseClass::OnAOEHitboxOverlapBegin);
 }
 
 void AMyProjectileBaseClass::Tick(float DeltaTime)
 {
 	Super::Tick(DeltaTime);
-}
\ No newline at end of file
+
+	if (FVector::Dist(StartingLocation, GetActorLocation()) > ProjectileRange)
+	{
+		this->Destroy();
+	}
+}
+
+void AMyProjectileBaseClass::OnHitboxOverlapBegin(UPrimitiveComponent* OverlappedComp, AActor* OtherActor, UPrimitiveComponent* OtherComp, int32 OtherBodyIndex, bool bFromSweep, const FHitResult& SweepResult)
+{
+	if (Enemy == OtherActor)
+	{
+		Enemy = Cast<AMyEnemyBaseClass>(OtherActor);
+
+		Enemy->HitEnemy(Damage);
+
+		switch (ProjectileType)
+		{
+		case EProjectileType::Physical:
+			this->Destroy();
+
+			break;
+
+		case EProjectileType::Fire:
+			Mesh->SetVisibility(false);
+			ProjectileHitbox->SetActive(false);
+			AreaOfEffectHitbox->SetActive(true);
+
+			break;
+
+		case EProjectileType::Water:
+			//add confusion here
+			this->Destroy();
+
+			break;
+
+		case EProjectileType::Poison:
+			Mesh->SetVisibility(false);
+			ProjectileHitbox->SetActive(false);
+
+			GetWorldTimerManager().SetTimer(DamageOverTimeTimerHandle, this, &AMyProjectileBaseClass::DamageOverTimeEnemy,
+				DamageOverTimeRate, true, DamageOverTimeRate);
+
+			break;
+
+		case EProjectileType::Ice:
+			Enemy->SetIsFrozen(true);
+			this->Destroy();
+
+			break;
+		}
+	}
+	else if (Steve == OtherActor)
+	{
+		Steve->HitPlayer(Damage);
+
+		switch (ProjectileType)
+		{
+		case EProjectileType::Physical:
+			this->Destroy();
+
+			break;
+
+		case EProjectileType::Fire:
+			Mesh->SetVisibility(false);
+			ProjectileHitbox->SetActive(false);
+			AreaOfEffectHitbox->SetActive(true);
+
+			break;
+
+		case EProjectileType::Water:
+			//add effect here
+			this->Destroy();
+
+			break;
+
+		case EProjectileType::Poison:
+			Mesh->SetVisibility(false);
+			ProjectileHitbox->SetActive(false);
+
+			GetWorldTimerManager().SetTimer(DamageOverTimeTimerHandle, this, &AMyProjectileBaseClass::DamageOverTimePlayer,
+				DamageOverTimeRate, true, DamageOverTimeRate);
+
+			break;
+
+		case EProjectileType::Ice:
+			//add effect here
+			this->Destroy();
+
+			break;
+		}
+	}
+	else
+	{
+		if (ProjectileType == EProjectileType::Fire)
+		{
+			Mesh->SetVisibility(false);
+			ProjectileHitbox->SetActive(false);
+			AreaOfEffectHitbox->SetActive(true);
+		}
+		else
+		{
+			this->Destroy();
+		}
+	}
+}
+
+void AMyProjectileBaseClass::OnAOEHitboxOverlapBegin(UPrimitiveComponent* OverlappedComp, AActor* OtherActor, UPrimitiveComponent* OtherComp, int32 OtherBodyIndex, bool bFromSweep, const FHitResult& SweepResult)
+{
+	if (Enemy == OtherActor)
+	{
```

Comment:
if (Enemy == OtherActor) could we write theses the other way?
if (OtherActor == Enemy) makes more sense to me this way..